qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -0,0 +1,235 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A generalized QAOA quantum circuit with a support of custom initial states and mixers."""
+# pylint: disable=R0401
+from typing import Optional, Union, cast, Set
+
+import numpy as np
+
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.parametervector import ParameterVector
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from ..blueprintcircuit import BlueprintCircuit
+
+
+class QAOAAnsatz(BlueprintCircuit):
+    """A generalized QAOA quantum circuit with a support of custom initial states and mixers.
+
+    References:
+
+        [1]: Farhi et al., A Quantum Approximate Optimization Algorithm.
+            `arXiv:1411.4028 <https://arxiv.org/pdf/1411.4028>`_
+    """
+    def __init__(self,
+                 cost_operator: 'OperatorBase',
```

Comment:
Is there a reason the type is given as a string. Normally its to cope with forward references. In this case can we not just import the type and use it directly rather than as a string?

Reply:
A direct import of `OperatorBase` causes a circular import problem, `qiskit.circuit.library` imports opflow and opflow imports `QuantumCircuit`. Moreorver, if I specify type hints as `'OperatorBase'`, sphinx fails to build the documentation, saying this type is not defined. I discussed this problem with @Cryoris, agreed to move type hints to docstrings. Not great, but hope this version can pass all checks.

Reply:
Ok, I agree its not great. The import situation ought to be looked at in the future to see if things can be altered to avoid that problem